maestro-web/src/main/kotlin/maestro/web/record/WebScreenRecorder.kt

```diff
@@ -0,0 +1,102 @@
+package maestro.web.record
+
+import okio.Sink
+import org.openqa.selenium.WebDriver
+import org.openqa.selenium.devtools.HasDevTools
+import org.openqa.selenium.devtools.v130.page.Page
+import java.util.*
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+
+class WebScreenRecorder(
+    private val videoEncoder: VideoEncoder,
+    private val seleniumDriver: WebDriver
+) : AutoCloseable {
+
+    private val screenRecordingSessions = mutableListOf<AutoCloseable>()
+    private lateinit var recordingExecutor: ExecutorService
+
+    private var closed = false
+
+    fun startScreenRecording(out: Sink) {
+        ensureNotClosed()
+
+        recordingExecutor = Executors.newSingleThreadExecutor()
+        videoEncoder.start(out)
+
+        startScreenRecordingForCurrentWindow()
+    }
+
+    fun onWindowChange() {
+        if (closed) {
+            return
+        }
+
+        startScreenRecordingForCurrentWindow()
+    }
+
+    override fun close() {
+        if (closed) {
+            return
+        }
+        closed = true
+
+        closeScreenRecordingSessions()
+
+        recordingExecutor.shutdown()
+        recordingExecutor.awaitTermination(2, TimeUnit.MINUTES)
+
+        videoEncoder.close()
+    }
+
+    private fun startScreenRecordingForCurrentWindow() {
+        closeScreenRecordingSessions()
+
+        val driver = seleniumDriver as HasDevTools
+
+        val seleniumDevTools = driver.devTools
+
+        seleniumDevTools.createSession()
+
+        seleniumDevTools.send(Page.enable())
+
+        seleniumDevTools.send(
+            Page.startScreencast(
+                Optional.of(Page.StartScreencastFormat.JPEG),
+                Optional.of(80),
+                Optional.of(1280),
+                Optional.of(1280),
```

Comment:
Will it still use the dimensions of the current window? Looks like it's the max width / height according to the API, but curious if we should instead pull the dimensions from the current window to pipe in here.

Reply:
Yes, these are indeed the max dimensions so the actual recording will keep the proper aspect ratio.

I was originally using the full device size, but this results in a substantially slower-to-process video at the end (since we are capturing quite a lot of frames), so I settled for some reasonable default.